### PR TITLE
Presheaf categories

### PR DESCRIPTION
**Pull request by fredrik-bakke**
### Summary
- Define commuting squares in large precategories
- Define maps from small to large (pre)categories
- Define functors from small to large (pre)categories
- Define (large) (pre)categories of functors/maps from small (pre)categories to large (pre)categories
- Define presheaf categories

I also had some other things lying around in the branch, sorry about that:
- Define representable functors on large precategories
- Show that functors on small (pre)categories preserve isomorphism
- Minor renamings

Completes item 2 and 19 of #769

### CODE REVIEW

File: src/category-theory/functors-categories.lagda.md [hunk 1]
```diff
@@ -39,16 +40,16 @@ module _
   (F : map-Category C D)
   where
 
-  preserves-comp-map-Category : UU (l1 ⊔ l2 ⊔ l4)
-  preserves-comp-map-Category =
-    preserves-comp-map-Precategory
+  preserves-comp-hom-map-Category : UU (l1 ⊔ l2 ⊔ l4)
```
Comment (fredrik-bakke) [inline src/category-theory/functors-categories.lagda.md:43]: I chose to rename this, as for maps between categories, one could ask that it for instance only preserves isomorphisms.

File: src/category-theory/functors-small-large-precategories.lagda.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+### Functors from small to large precategories
+
+```agda
+module _
+  {l1 l2 : Level} {α : Level → Level} {β : Level → Level → Level}
+  (C : Precategory l1 l2) (D : Large-Precategory α β)
+  where
+
+  functor-Small-Large-Precategory : (γ : Level) → UU (l1 ⊔ l2 ⊔ α γ ⊔ β γ γ)
```
Comment (fredrik-bakke) [inline src/category-theory/functors-small-large-precategories.lagda.md:86]: Would you prefer `functor-Precategory-Large-Precategory`?

File: src/category-theory/functors-small-large-precategories.lagda.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  functor-Small-Large-Precategory γ =
+    Σ ( obj-Precategory C → obj-Large-Precategory D γ)
```
Comment (EgbertRijke) [inline src/category-theory/functors-small-large-precategories.lagda.md:88]: Should this be defined as `functor-Precategory C (precategory-Large-Precategory D gamma)`?

Comment (fredrik-bakke) [reply src/category-theory/functors-small-large-precategories.lagda.md:88]: ah, yeah, that works of course!

Comment (fredrik-bakke) [reply src/category-theory/functors-small-large-precategories.lagda.md]: For posterity, the short answer is: "`Level` shouldn't be a treated as a type"

**Comment by fredrik-bakke** [comment]
Just a heads-up, I'm going straight for presheaf categories now @maybemabeline

**Comment by fredrik-bakke** [comment]
I think it is best if this PR is merged once I have the definitions written down. After that, you can create a new PR doing what you want with them. Currently, I think I am content with just defining them, although that can of course change 🙃

File: src/category-theory/functors-small-large-categories.lagda.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
+# Functors from small to large categories
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/category-theory/functors-small-large-categories.lagda.md:1]: sure thing

File: src/category-theory/category-of-functors-small-large-categories.lagda.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+[categories](category-theory.categories.md) to
+[large categories](category-theory.large-categories.md) and
+[natural transformations](category-theory.natural-transformations-functors-small-large-precategories.md)
+between them form a large category whose identity map and composition structure
+are inherited pointwise from the codomain category. This is called the
+**category of functors from small to large categories**.
+
+## Lemmas
+
+### Extensionality of functors from small to large precategories when the codomain is a large category
```
Comment (EgbertRijke) [inline src/category-theory/category-of-functors-small-large-categories.lagda.md:42]: Why not: Extensionality of functors from small precategories to large categories?

File: src/category-theory/category-of-functors-small-large-categories.lagda.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+```agda
+module _
+  {l1 l2 : Level} {α : Level → Level} {β : Level → Level → Level}
+  (C : Precategory l1 l2)
+  (D : Large-Precategory α β)
+  (is-large-category-D : is-large-category-Large-Precategory D)
+  where
+
+  equiv-natural-isomorphism-htpy-functor-is-category-Small-Large-Precategory :
```
Comment (EgbertRijke) [inline src/category-theory/category-of-functors-small-large-categories.lagda.md:52]: The predicate is called `is-large-category` so I'm not sure if using `is-category` in this name is a good idea. For consistency, in other instances I've always put `is-large-category` in the name if `is-large-category` is assumed somewhere.

File: src/category-theory/category-of-functors-small-large-categories.lagda.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+## Lemmas
+
+### Extensionality of functors from small to large precategories when the codomain is a large category
+
+```agda
+module _
+  {l1 l2 : Level} {α : Level → Level} {β : Level → Level → Level}
+  (C : Precategory l1 l2)
+  (D : Large-Precategory α β)
+  (is-large-category-D : is-large-category-Large-Precategory D)
```
[the remark this answers is not in the document]
Comment (fredrik-bakke) [reply src/category-theory/category-of-functors-small-large-categories.lagda.md:49]: We had a discussion about this the other day, and this way there is a clear name for the entry, while the suffix also makes sense

[the remark this answers is not in the document]
Comment (EgbertRijke) [reply src/category-theory/category-of-functors-small-large-categories.lagda.md:49]: Ok, sounds good.

**Comment by EgbertRijke** [comment]
This is quite amazing! I can't believe how fast you're adding stuff to the category theory folder these days 😁

**Comment by EgbertRijke** [comment]
Ah, but it is still in draft. How much do you still want to add??:)

**Comment by fredrik-bakke** [comment]
> This is quite amazing! I can't believe how fast you're adding stuff to the category theory folder these days 😁

Thanks, that means a lot coming from you 🥹

**Comment by fredrik-bakke** [comment]
By the way @maybemabeline, were you planning on adding representable presheaves? Otherwise, I felt a sudden inspiration to add those as well (in my next PR)

**Comment by fredrik-bakke** [comment]
Ah, there's a slight kink in the naming now. We have the file named `natural-transformations-from-small-to-large-precategories`. Should we rename our natural transformations files to specify that they are transformations between functors?

**Comment by EgbertRijke** [comment]
`natural-transformations-functors-from-small-to-large-precategories`

(I would omit "between")

**Review by fredrik-bakke — COMMENTED** [review]
Alright, I'm happy with it now I think

**Pull request merged**